packages/firestore/src/core/database_info.ts

```diff
@@ -77,6 +77,10 @@ export class DatabaseId {
       other.database === this.database
     );
   }
+
+  canonicalString(): string {
```

Comment:
After a lot of consideration, I think you should get rid of this "canonicalString()" method and simply inline its implementation into the single place that it's used. IMO This little bit of abstraction detracts from readability at the call site.